### DISCUSSION TITLE
Polymorphic default value on abstract type

### DISCUSSION THREAD

**Discussion by rdlaitila** [geldata/gel#3394] [Q&A] [answered]
I have a use case which requires type information to be provided on the returning database objects. However i'm struggling to find any syntax that allows this to be expressed:

```edgeql
module default {
	abstract type Entity {
		required property etype -> str {
			default := (
				# Possible to assign type string from the concrete type being constructed?
				# ie: [is Account]"entity:type:Account"
				# ie: [is Event]"entity:type:Event"
				# ie: [is User]"entity:type:User"
			)
		};
	}

	type Account extending Entity {			
	}

	type Event extending Entity {				
	}

	type User extending Entity {				
	}
}
```

I can place a computed property for the `etype` directly on the concrete types, but when selecting links containing just the abstract `default::Entity` I cannot shape the `etype` without placing type switches for each entity in the system for each select query involving `default::Entity`:

```
module default {
	abstract type Entity {		
	}

	type Account extending Entity {	
		required property etype := "entity:type:Account";
	}

	type Event extending Entity {				
		required property etype := "entity:type:Event";
	}

	type User extending Entity {				
		required property etype := "entity:type:User";
	}
}
```

```
select default::Entity {
	[is default::Account].etype,
	[is default::Event].etype,
	[is default::User].etype,
}
```

Any thoughts on possible solutions?

**Comment by elprans** [comment, marked as answer]
> I have a use case which requires type information to be provided on the returning database objects

Every object already has a [`__type__` link](https://www.edgedb.com/docs/guides/introspection/index) which points to it's reflected type in `schema::ObjectType`, so you normally don't need to roll your own:

```
select default::Entity {
   etype := .__type__.name
}
```

if you absolutely need a custom naming scheme, you can do something like this:

```
abstract type Entity {	
   property etype := "entity:type:" ++ str_split(.__type__.name, "::")[1]
}
```

**Reply by rdlaitila** [reply]
Thanks for the help I was dancing around the `.__type__` link but it makes perfect sense now!
